### PR TITLE
[FrameworkBundle] Add integration of http-client component

### PR DESCRIPTION
**Pull request by Ioni14**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This PR adds the integration of the HttpClient component on FrameworkBundle.
By default, two services are provided, one implementing SFC-HttpClient, and another PSR18:
* `http_client` + its autowiring alias for `Symfony\Contracts\HttpClient\HttpClientInterface`)
This service is automatically set to the best HTTP client available with the configuration given under the `framework.http_client` key.
* `psr18.http_client` + its autowiring alias for `Psr\Http\Client\ClientInterface`). To make it work, one needs to provide autowiring aliases for `ResponseFactoryInterface` and `StreamFactoryInterface`, which are provided by [the recipe](https://github.com/symfony/recipes-contrib/blob/master/nyholm/psr7/1.0/config/packages/nyholm_psr7.yaml) for `nyholm/psr7` (but could be overriden by apps when using something else).

* one can also configure the default options, and "scoped" clients. For example:
```yaml
http_client:
    default_options:
        capath: '...'
    clients:
        github_client:
            default_options:
                base_uri: 'https://api.github.com'
```

This definition create a `github_client` service implementing SFC-HttpClient and a `psr18.github_client` one implementing PSR18, +2 corresponding named autowiring aliases: `HttpClientInterface $githubClient`,  and `ClientInterface $githubClient`.

### CODE REVIEW

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php
Comment (linaori) [inline src/Symfony/Component/HttpClient/Response/CurlResponse.php]: This is an odd if, could be changed into a guard clause with early return

File: src/Symfony/Component/HttpClient/Response/NativeResponse.php
Comment (linaori) [inline src/Symfony/Component/HttpClient/Response/NativeResponse.php]: As this is internal, why not move the int to the typehints? Should result in the same behavior and Symfony can regulate the strictness of this call

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php
Comment (Ioni14) [reply src/Symfony/Component/HttpClient/Response/CurlResponse.php]: Hello, 
You should review this PR if it's about the component itself. ;)
https://github.com/symfony/symfony/pull/30413

Comment (linaori) [reply src/Symfony/Component/HttpClient/Response/CurlResponse.php]: I can't tell if your changes are a PR to the same branch including the other changes. At this point it's impossible to properly review until that one is merged.

@nicolas-grekas, as you're the author of that PR, this comment and the one below are probably for you :smile:

**Comment by javiereguiluz** [comment]
@Ioni14 thanks for your hard work on this feature!

A minor question: could we maintain the global `default_options` option ... but remove it from the individual HTTP clients? In the example you showed in the description, we'd maintain the first `default_options` but we'd remove the other two. Don't they look a bit redundant? Thanks.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Should we put the default values here for the options that define them? For example, `20` for the `max_reidrects`, `true` or `false` for `buffer`, etc.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `Invalid API client ...` ->  `Invalid HTTP client ...`

**Comment by Ioni14** [comment]
@javiereguiluz thanks ;)

It's probably "YAGNI" but if we want to add other configuration keys for clients in the future it'll be useful to have `default_options`.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd
Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: should we sort these options alphabetically? I don't know what's the Symfony policy about this in XSD files.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (sstok) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Can't we use `append()` here? 

And why is there both `http_client` and `http_clients`? If the purpose is to have a default maybe add a `default: true` key for `http_clients`, but warn if this is used more then once 🤔 or `default_http_client` (al though a bit of context).

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (derrabus) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: The component already defines defaults for those options. Do we need different ones for the bundle?

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: I agree with @derrabus

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: Keeping the order from HttpClientInterface looks consistent to me

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: fixed

**Comment by nicolas-grekas** [comment]
PR and description updated, ready.

**Review by nicolas-grekas — APPROVED** [review]

**Review by derrabus — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: I wouldn't return a node builder here. This just makes it harder to follow the other tree modifications in `addHttpClientSection()`.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: If I don't miss anything, this cannot be set when using the XML config  format.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: `integerNode()`?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: `xsd:integer`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: I think we should allow `header` elements in the XML format (thus need `fixXmlConfig('header')`)

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Should we reuse `$config['max_host_connections']` to not have to configure this value for each client?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Do we really want to use autowiring here? IIRC we don't do that anywhere else with core services.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/http_client.xml
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/http_client.xml]: Do we really want to use autowiring here?

File: src/Symfony/Component/HttpClient/CurlHttpClient.php
Comment (xabbuh) [inline src/Symfony/Component/HttpClient/CurlHttpClient.php]: `(int)`?

File: src/Symfony/Component/HttpClient/NativeHttpClient.php
Comment (xabbuh) [inline src/Symfony/Component/HttpClient/NativeHttpClient.php]: `(int)`?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: fixed

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: not sure what you missed, I added it in the fixtures and it worked - or did you mean something else? (see fixtures)

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: timeout is a float: 0.1 means 100ms

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: on it

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: done

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: autowiring allow providing working services once the required aliases are installed
see eg https://github.com/symfony/recipes/blob/master/nyholm/psr7/1.0/config/packages/nyholm_psr7.yaml
This looked the best to achieve loose coupling at the configuration level to me. WDYT?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/http_client.xml
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/http_client.xml]: see above

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: just noticed: we should use dashes every in XML instead of underscores

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: I think we should auto-enable this when the `http-client` component is installed. That's consistent with other components, and makes it so that, *just* by forcing the user to install `http-client`, we can more-or-less safely know that the `http_client` service

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (weaverryan) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Maybe related to this, I'm currently getting a fatal error when running `debug:container` or `debug:config` with this component enabled in the config:

```
PHP Fatal error:  Method class@anonymous::__toString() must not throw an exception, caught Symfony\Component\DependencyInjection\Exception\AutowiringFailedException:  in /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php on line 229
PHP Stack trace:
PHP   1. {main}() /Users/weaverryan/Sites/os/docs/http_client_play/bin/console:0
PHP   2. Symfony\Bundle\FrameworkBundle\Console\Application->run() /Users/weaverryan/Sites/os/docs/http_client_play/bin/console:38
PHP   3. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/Console/Application.php:145
PHP   4. Symfony\Bundle\FrameworkBundle\Console\Application->doRun() /Users/weaverryan/Sites/os/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:73
PHP   5. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/Console/Application.php:269
PHP   6. Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() /Users/weaverryan/Sites/os/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:87
PHP   7. Symfony\Bundle\FrameworkBundle\Command\ContainerDebugCommand->run() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/Console/Application.php:926
PHP   8. Symfony\Bundle\FrameworkBundle\Command\ContainerDebugCommand->execute() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/Console/Command/Command.php:255
PHP   9. Symfony\Bundle\FrameworkBundle\Command\ContainerDebugCommand->getContainerBuilder() /Users/weaverryan/Sites/os/symfony/src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php:117
PHP  10. Symfony\Component\DependencyInjection\ContainerBuilder->compile() /Users/weaverryan/Sites/os/symfony/src/Symfony/Bundle/FrameworkBundle/Command/ContainerDebugCommand.php:213
PHP  11. Symfony\Component\DependencyInjection\Compiler\Compiler->compile() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/ContainerBuilder.php:754
PHP  12. Symfony\Component\DependencyInjection\Compiler\AutowirePass->process() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/Compiler.php:95
PHP  13. Symfony\Component\DependencyInjection\Compiler\AutowirePass->process() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:52
PHP  14. Symfony\Component\DependencyInjection\Compiler\AutowirePass->processValue() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AbstractRecursivePass.php:46
PHP  15. Symfony\Component\DependencyInjection\Compiler\AutowirePass->doProcessValue() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:71
PHP  16. Symfony\Component\DependencyInjection\Compiler\AutowirePass->processValue() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:99
PHP  17. Symfony\Component\DependencyInjection\Compiler\AutowirePass->processValue() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AbstractRecursivePass.php:82
PHP  18. Symfony\Component\DependencyInjection\Compiler\AutowirePass->doProcessValue() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:71
PHP  19. Symfony\Component\DependencyInjection\Compiler\AutowirePass->autowireCalls() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:122
PHP  20. Symfony\Component\DependencyInjection\Compiler\AutowirePass->autowireMethod() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:164
PHP  21. Symfony\Component\DependencyInjection\Compiler\AutowirePass->Symfony\Component\DependencyInjection\Compiler\{closure}() /Users/weaverryan/Sites/os/symfony/src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php:255
```

This is coming from the `psr18.http_client` service - not sure what the problem is though!

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: These are the default options for each request that's made. What about `default_request_options` (it's even documented as `Default requests' options` in `HttpClient` phpdoc). And also, should we make these options *not* automatically be used as the "default options" for all the other HTTP clients?

The problem is that there are 2 concepts conflicting right now. Does "default" mean:
A) default options for the request that can be overridden on each request
OR
B) default options that will be applied to all other HTTP clients

Right now, it's both - that confused me. It's really (A), so maybe we should make it *only* (A) and clarify the name.

Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: right now, it is default options applied to all other HTTP clients for their default request options...

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/http_client_full_default_options.xml
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/http_client_full_default_options.xml]: this is wrong. `foo` and `bar` cannot be part of the XSD here. As this is a prototyped array node, the key must belong to an attribute (having multiple nodes with a name)

And the same applies to other places

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/http_client_full_default_options.xml
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/http_client_full_default_options.xml]: this should be made as `<framework:header name="X-powered">PHP</framework:header>`

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/http_client_full_default_options.xml
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/http_client_full_default_options.xml]: these should use attributes rather than child tags, to be consistent with other parts of the config

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Correct, it's A and B. If one want to configure the default request options for the default client only, one should set them under `framework.http_cllient.clients.http_client.default_options` (since `http_client` is the name of the service that is aliased for `HttpClientInterface`).
I'm not sure I understand how your suggestion would apply. To me, all is fine :)

**Comment by weaverryan** [comment]
If/when #30547 is merged, the `auth` option in the config will need to change and the bearer_auth will need to be added.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (weaverryan) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: I was suggesting that we remove (B), but it's subjective. The question is: is it valuable for these options to cascade onto other clients? Or is it annoying/surprising, and then I am unsetting options on my individual clients, because I don't want them to cascade from my default client?

I'm not convinced that re-using options across clients will be all that common - things like auth, base_uri, headers all seem *very* client-specific. I could argue that it might be surprising if my auth on my default client starts getting sent to another server via another client (because I didn't realize the options would cascade down to my client).

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: e.g. cabundle/capath: you want to configure them once for all

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (weaverryan) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: See #30571

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/http_client_full_default_options.xml
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Fixtures/xml/http_client_full_default_options.xml]: one attribute per line would be great

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: since we restrict here, we should also restrict in Configuration, using an EnumNode

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (weaverryan) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Hmm, ok, really fine by me then.

**Review by weaverryan — APPROVED** [review]

**Comment by weaverryan** [comment]
Tests failing on low deps, but other than that - +1 from me!

**Comment by nicolas-grekas** [comment]
Thank you @Ioni14.

**Comment by nicolas-grekas** [comment]
Thank you @nicoweb

**Pull request merged**
